Add find_in_set Spark function

fyp711 opened the pull request:
Returns 1-based index of the given string in the comma-delimited list.
Returns 0, if the string was not found or if the given string contains a comma. 

Examples:

```
SELECT find_in_set('ab','abc,b,ab,c,def'); -- 3
SELECT find_in_set(NULL,'abc,b,ab,c,def'); -- NULL
SELECT find_in_set('abcdef','abc,b,ab,c,def'); -- 0
```

Spark documentation: https://spark.apache.org/docs/latest/api/sql/index.html#find_in_set

fyp711 commented:
I just modified the code logic to make it the same as Spark function logic

fyp711 commented:
Hi, I fixed the UT issue, could anyone help me re run UT ? Thanks.

PHILO-HE commented on velox/functions/sparksql/String.h:
```diff
@@ -1130,4 +1130,59 @@ struct ReplaceFunction {
+    int32_t index = 1;
+    int32_t lastComma = -1;
+
+    auto _match = match.str();
```
It doesn't satisfy the naming convention.
See
https://github.com/facebookincubator/velox/blob/main/CODING_STYLE.md#naming-conventions

PHILO-HE commented on velox/functions/sparksql/tests/StringTest.cpp:
```diff
@@ -813,5 +819,20 @@ TEST_F(StringTest, replace) {
+
+TEST_F(StringTest, findInSet) {
+  EXPECT_EQ(findInSet("ab", "abc,b,ab,c,def"), 3);
+  EXPECT_EQ(findInSet("abc", "abc,b,ab,c,def"), 1);
```
Is unicode considered in the implementation? I'm assuming the spark function supports it. Let's add tests for unicode input.

jinchengchenghh commented:
Please update the PR description as this PR https://github.com/facebookincubator/velox/pull/8356

jinchengchenghh commented on velox/docs/functions/spark/string.rst:
```diff
@@ -286,4 +286,12 @@ Unless specified otherwise, all functions return NULL if at least one of the arg
\ No newline at end of file
+        SELECT upper('SparkSql'); -- SPARKSQL
+
+.. spark:function:: find_in_set(str, str_array) -> integer
```
str_array->strArray

jinchengchenghh commented on velox/docs/functions/spark/string.rst:
````diff
@@ -286,4 +286,12 @@ Unless specified otherwise, all functions return NULL if at least one of the arg
+    Returns the index (1-based) of the given string ```str``` in the comma-delimited list ```str_array```.
+    Returns 0, if the string was not found or if the given string ```str``` contains a comma. ::
+
+        SELECT find_in_set('ab','abc,b,ab,c,def'); -- 3
````
Add more corner cases to example.

jinchengchenghh commented on velox/functions/sparksql/String.h:
```diff
@@ -1130,4 +1130,59 @@ struct ReplaceFunction {
+
+    for (int i = 0; i < _strArraySize; i++) {
+      if (_strArrayData[i] == ',') {
+        auto substr = _strArray.substr(lastComma + 1, i - (lastComma + 1));
```
`substr` will generate a lot of small strings, the number depends on which depends on the number of `,` , maybe https://[cplusplus.com/reference/string/string/compare/](https://cplusplus.com/reference/string/string/compare/) can help you.

fyp711 replied:
Fixed, thanks

fyp711 commented:
Hi @jinchengchenghh @PHILO-HE, Thank you very much for helping me review, I will fix it. thanks for your guidance.

fyp711 commented:
Hi @PHILO-HE @jinchengchenghh , Could you help me review again? I optimized the code to add some new cases. Thanks.

fyp711 commented on velox/functions/sparksql/String.h:
```diff
@@ -1130,4 +1130,59 @@ struct ReplaceFunction {
+    return doCall(result, str, strArray);
+  }
+
+  FOLLY_ALWAYS_INLINE bool
```
Fixed, thanks

fyp711 commented on velox/functions/sparksql/String.h:
```diff
@@ -1130,4 +1130,59 @@ struct ReplaceFunction {
+struct FindInSetFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE bool callNullFree(
```
Thanks you for review, `callNullFree` is now used, because if any of the arguments are null, `NULL` must be returned.

jinchengchenghh commented on velox/functions/sparksql/String.h:
```diff
@@ -1130,4 +1130,60 @@ struct ReplaceFunction {
+    auto matchString = match.str();
+    auto strArrayString = strArray.str();
+
+    auto found = matchString.find(',');
```
The `found` used only once, no need to define it.

fyp711 commented:
> The `found` used only once, no need to define it.

Hi @jinchengchenghh , it fixed, thank you for helping me review.

PHILO-HE commented on velox/docs/functions/spark/string.rst:
````diff
@@ -55,6 +55,18 @@ Unless specified otherwise, all functions return NULL if at least one of the arg
 
+.. spark:function:: find_in_set(str, strArray) -> integer
+
+    Returns the index (1-based) of the given string ```str``` in the comma-delimited list ```strArray```.
````
Nit: \``str\``. Ditto for others.

PHILO-HE reviewed (commented):
Could you add a link to spark doc for this function in pr description? Thanks!

fyp711 commented:
> Could you add a link to spark doc for this function in pr description? Thanks!

@PHILO-HE I have added the document of spark function in the description, and the issues mentioned above have been fixed. I also added test cases for non-ASCLL characters. Let's wait the CI success. Could you help me to check what needs to be improved, and I will fix it as soon as possible. Thanks for helping me reivew the code.

fyp711 commented:
Hi everyone, all the CI tests passed.

PHILO-HE commented on velox/functions/sparksql/tests/StringTest.cpp:
```diff
@@ -813,5 +819,43 @@ TEST_F(StringTest, replace) {
+  EXPECT_EQ(
+      findInSet(
+          "\u0061\u0062",
+          "\u0061\u0062\u0063\u002c\u0062\u002c\u0061\u0062\u002c\u0063\u002c\u0064\u0065\u0066"),
```
Since those are all ascii characters. Let's directly use the corresponding character instead of code point for readability. Ditto for other applicable places.

PHILO-HE reviewed (commented):
Looks good! Some trivial comments.

PHILO-HE commented on velox/functions/sparksql/tests/StringTest.cpp:
```diff
@@ -813,5 +819,34 @@ TEST_F(StringTest, replace) {
 }
+
+TEST_F(StringTest, findInSet) {
+  EXPECT_EQ(findInSet("ab", "abc,b,ab,c,def"), 3);
```
Can `strArray` have same elements? Thanks!

fyp711 commented:
> Can strArray have same elements? Thanks!
@PHILO-HE Yes, it will return the first index that appears, and I'll add some test cases later.

fyp711 commented:
> Looks good! Some trivial comments.

@PHILO-HE Let me refine the code. Your review has helped me a lot, thank you very much.

fyp711 commented on velox/functions/sparksql/String.h:
```diff
@@ -1130,4 +1130,53 @@ struct ReplaceFunction {
+struct FindInSetFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE bool call(
```
I will fix it.

fyp711 commented on velox/functions/sparksql/tests/StringTest.cpp:
```diff
@@ -813,5 +819,43 @@ TEST_F(StringTest, replace) {
+TEST_F(StringTest, findInSet) {
+  EXPECT_EQ(findInSet("ab", "abc,b,ab,c,def"), 3);
+  EXPECT_EQ(findInSet("abc", "abc,b,ab,c,def"), 1);
+  EXPECT_EQ(findInSet("ab", "abc,b,ab,ab,ab"), 3);
```
If `strArray` have same elements, it will return the first index that appers.

PHILO-HE commented on velox/functions/sparksql/String.h:
```diff
@@ -1130,4 +1130,53 @@ struct ReplaceFunction {
+    int32_t index = 1;
+    int32_t lastComma = -1;
+
+    auto strString = std::string_view(str);
```
Maybe, better to rename it, as well as `strArrayString`. Thanks!

fyp711 replied:
Ok, The strString will be renamed to `match` and the strArrayString is `arrayStr`. Do you think arrayStr is good ?Thanks.

PHILO-HE commented on velox/functions/sparksql/String.h:
```diff
@@ -1130,4 +1130,52 @@ struct ReplaceFunction {
+    int32_t lastComma = -1;
+
+    auto match = std::string_view(str);
+    auto arrayStr = std::string_view(strArray);
```
I think we can directly use `strArray` to call `compare()`. It's `StringView`. If so, we can remove this line.

fyp711 replied:
The StringView compare method takes only one parameter for comparison, which is not what I want.

fyp711 replied:
> The StringView compare method takes only one parameter for comparison, which is not what I want.

I need to compare characters of the specified length in `strArray`, such as `find_in_set("d","abc,d,e,f")`, I need to pick up `abc` to copare with `d`, and then pick `d` to compare with `str`, and cycle like this. So The StringView's compare method is doesn't meet my needs.

PHILO-HE replied:
One way is to construct a StringView for a substring, e.g, `StringView(strArray.data() + offset, len)`, then we can directly use "==" operator to compare it with `str`. `StringView`'s compare() is also good, but looks "==" is more straightforward.
https://github.com/facebookincubator/velox/blob/main/velox/type/StringView.h#L134

fyp711 replied:
> One way is to construct a StringView for a substring, e.g, `StringView(strArray.data() + offset, len)`, then we can directly use "==" operator to compare it with `str`. `StringView`'s compare() is also good, but looks "==" is more straightforward. 

Thanks, I think this  will generates a lot of substrings, it's depend on the number of `,`,  If so, it's not the best way. What do you think？

fyp711 replied:
> Thanks, I think this will generates a lot of substrings, it's depend on the number of `,`, If so, it's not the best way. What do you think？
It seems that we can avoid this issue, let me implement it first.Thank you for review.

PHILO-HE replied:
It will not copy the substring, but only creates some extra StringView instances. 
I note StringView is velox-optimized for better comparison. I'm not sure how much performance gain it will bring out. Maybe, you can just use any one (StringView or std::string_view for comparing) and let masha decide in final review. Thanks!

fyp711 replied:
Ok, thanks. I'm going to try to optimize with StringView.

fyp711 replied:
I've already used StringView, I need to test it, wait CI success . Thanks

fyp711 replied:
All CI tests passed.

PHILO-HE commented on velox/functions/sparksql/String.h:
```diff
@@ -1130,4 +1130,49 @@ struct ReplaceFunction {
+    int32_t index = 1;
+    int32_t lastComma = -1;
+
+    if (str.find(',') != std::string::npos) {
```
should be `std::string_view(str).find(..`?

fyp711 replied:
yes, it will be fixed a moment

fyp711 commented:
@PHILO-HE Could you help review again. Thanks.

PHILO-HE commented on velox/functions/sparksql/String.h:
```diff
@@ -1130,4 +1130,52 @@ struct ReplaceFunction {
+    size_t arraySize = strArray.size();
+    size_t matchSize = str.size();
+
+    for (int i = 0; i < arraySize; i++) {
```
One possible optimization.
For constant `strArray`, in `initialize()` (see [link](https://github.com/facebookincubator/velox/blob/main/velox/docs/develop/scalar-functions.rst#access-to-session-properties-and-constant-inputs)), I think we can split it with comma delimiter and create a hash map to hold unique elements and their indexes. Then, in `call()`, for such constant case, we can just do a search from hash map for each row. Does it make sense?

For non-constant case, I think we can keep the current handling unchanged in `call`.

fyp711 replied:
Thanks, I think both `str` and `strArray` might be a different value, i think that we can keep our current processing logic.

PHILO-HE replied:
Not sure how common column `strArrray` is used.
Hi @mbasmanova, I have reviewed this pr. Do you have any comment on this or other part? Thanks!

mbasmanova commented on velox/docs/functions/spark/string.rst:
```diff
@@ -55,6 +55,18 @@ Unless specified otherwise, all functions return NULL if at least one of the arg
 
+.. spark:function:: find_in_set(str, strArray) -> integer
+
+    Returns the index (1-based) of the given string ``str`` in the comma-delimited list ``strArray``.
```
nit: Returns 1-based index ...

fyp711 replied:
Fixed, Thanks!

mbasmanova commented on velox/docs/functions/spark/string.rst:
```diff
@@ -55,6 +55,18 @@ Unless specified otherwise, all functions return NULL if at least one of the arg
+    Returns 0, if the string was not found or if the given string ``str`` contains a comma. ::
+
+        SELECT find_in_set('ab', 'abc,b,ab,c,def'); -- 3
+        SELECT find_in_set('', ''); -- 1
```
Would you add examples where return value is 0, both not-found and str-contains-comma?

fyp711 replied:
Fixed, Thanks!

mbasmanova commented on velox/functions/sparksql/String.h:
```diff
@@ -1130,4 +1130,52 @@ struct ReplaceFunction {
+      out_type<int32_t>& result,
+      const arg_type<Varchar>& str,
+      const arg_type<Varchar>& strArray) {
+    int32_t index = 1;
```
move these variables after the check for whether 'str' contains a comma

fyp711 replied:
Thanks. I will fix it.

mbasmanova commented on velox/functions/sparksql/String.h:
```diff
@@ -1130,4 +1130,52 @@ struct ReplaceFunction {
+    for (int i = 0; i < arraySize; i++) {
+      if (arrayData[i] == ',') {
+        if (i - (lastComma + 1) == matchSize &&
+            str ==
```
why not use std::memcmp?

fyp711 replied:
Thank you for reivew, This is a good choice, The performance should be better using memcmp, and I will change to this way in a moment.I also did some tests and the results showed that std::memcmp performed better.Here are some of my test results.

`test group 1`
memcmp cost time : 	8.2309s
memcmp cost time : 	8.12958s
memcmp cost time : 	8.08002s
compare cost time : 	9.72118s
compare cost time : 	9.80633s
compare cost time : 	9.95188s

`test group 2`
memcmp cost time : 	15.6135s
memcmp cost time : 	15.7243s
memcmp cost time : 	15.6123s
compare cost time : 	18.8385s
compare cost time : 	19.0486s
compare cost time : 	19.1677s

mbasmanova commented on velox/functions/sparksql/String.h:
```diff
@@ -1130,4 +1130,52 @@ struct ReplaceFunction {
+    size_t arraySize = strArray.size();
+    size_t matchSize = str.size();
+
+    for (int i = 0; i < arraySize; i++) {
```
why not use std::string_view::find?

https://en.cppreference.com/w/cpp/string/basic_string_view/find

fyp711 replied:
If this is the case, the find method may not find the correct query
`find_in_set("ab","abcdefg,ab,hh")`

mbasmanova replied:
I mean why not use std::string_view::find to find next comma.

fyp711 replied:
I know what you mean, but I'm not sure which of the two approaches is better, The reason I chose this approach is to keep it consistent with the spark function. Thanks.

mbasmanova replied:
I suggest to write a small benchmark to see if there is any perf difference.

See velox/benchmarks/basic/LikeBenchmark.cpp for an example.

fyp711 replied:
see https://github.com/apache/spark/blob/667c0a9dbbe045c73842a345c1b3897b155564d4/common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java#L471

fyp711 replied:
I think we should keep current logic, I did three sets of tests, and I found that the while loop way performed better than the `std::string_view::find` method, Here are some of my test results.` test goup1` tests to find the substring at the front of the strArray,` test group 2` at the middle, and `test goup 3` at the end. As the substring appears further back, the performance of the method using string_view::find decreases more. Thanks.

`test group 1`:
while loop cost time : 	5.7535s
while loop cost time : 	5.97309s
while loop cost time : 	5.71566s
string_view::find  cost time : 	5.96654s
string_view::find  cost time : 	5.97334s
string_view::find  cost time : 	5.97478s


`test group test2`:
while loop cost time : 	9.72591s
while loop cost time : 	9.78818s
while loop cost time : 	9.76955s
string_view::find  cost time : 	14.807s
string_view::find  cost time : 	14.8721s
string_view::find  cost time : 	14.785s

`test group test3`:
while loop cost time : 	19.2678s
while loop cost time : 	18.9274s
while loop cost time : 	18.7739s
string_view::find  cost time : 	27.0144s
string_view::find  cost time : 	27.2736s
string_view::find  cost time : 	27.4933s

fyp711 replied:
In my tests, `string_view::find` performs better when determining whether a substring contains a comma, and the for loop performs better when finding comma indexs. I think this is the best combination. Thanks!

mbasmanova replied:
Got it. Thank you for checking.

mbasmanova commented on velox/docs/functions/spark/string.rst:
```diff
@@ -55,6 +55,18 @@ Unless specified otherwise, all functions return NULL if at least one of the arg
+    Returns 0, if the string was not found or if the given string ``str`` contains a comma. ::
+
+        SELECT find_in_set('ab', 'abc,b,ab,c,def'); -- 3
+        SELECT find_in_set('', ''); -- 1
```
> SELECT find_in_set('', ''); -- 1

This behavior is surprising. Perhaps, document it explicitly. Is there a unit test to verify this?

mbasmanova commented on velox/functions/sparksql/String.h:
```diff
@@ -1130,4 +1130,52 @@ struct ReplaceFunction {
+        if (i - (lastComma + 1) == matchSize &&
+            str ==
+                StringView(
+                    strArray.data() + (lastComma + 1), i - (lastComma + 1))) {
```
strArray.data() -> arrayData
i - (lastComma + 1) -> matchSize

fyp711 replied:
I will fix it. Thanks！

mbasmanova reviewed (commented):
Overall looks good % some comments.

fyp711 commented on velox/functions/sparksql/tests/StringTest.cpp:
```diff
@@ -813,5 +819,35 @@ TEST_F(StringTest, replace) {
+  EXPECT_EQ(findInSet("dfg", "abc,b,ab,c,def"), 0);
+  EXPECT_EQ(findInSet("dfg", "dfgdsiaq"), 0);
+  EXPECT_EQ(findInSet("dfg", "dfgdsiaq, dshadad"), 0);
+  EXPECT_EQ(findInSet("", ""), 1);
```
> This behavior is surprising. Perhaps, document it explicitly. Is there a unit test to verify this?

@mbasmanova  I've added this test.

mbasmanova reviewed (approved):
Thanks.
